### DISCUSSION TITLE
Adding stimulus duration to moving-window plugin

### DISCUSSION THREAD

**Discussion by emilymech** [jspsych/jsPsych#998] [Q&A] [answered]
Hi everyone!

I'm working on a self-paced reading experiment, and I've adapted the moving-window plugin from the moving research online tutorial (thanks!!). I'm presenting words in the middle of the screen, and my edit to accomplish that is working well. However, I would like to add a parameter that will limit the stimulus duration (a word) to whatever duration the user sets. I've tried to adapt the plugin to be similar to another plugin that has a stimulus_duration parameter, but it's not working as of yet in my own experiment.

Here's is the code for the plugin as it currently is:
```
/*
 * Example plugin template
 */

jsPsych.plugins["moving-window"] = (function() {

  var plugin = {};

  plugin.info = {
    name: "moving-window",
    parameters: {
      words: {
        type: jsPsych.plugins.parameterType.STRING, // BOOL, STRING, INT, FLOAT, FUNCTION, KEYCODE, SELECT, HTML_STRING, IMAGE, AUDIO, VIDEO, OBJECT, COMPLEX
        default: undefined
      },
      key: {
        type: jsPsych.plugins.parameterType.KEYCODE,
        default: 32
      },
      stimulus_duration:{        
        type: jsPsych.plugins.parameterType.INT,
        pretty_name: 'Stimulus duration',
        default: null,
        description: 'How long to hide the stimulus.'
      }
    }
  }

  plugin.trial = function(display_element, trial) {

    // initalize some variables
    var trial_data = { words: trial.words }; // data object for the trial
    var n_words = trial.words.split(' ').length; // number of words in the trial
    var rt = []; // empty array for collecting RTs
    var current_word = 0; // current word

    // create a function for generating the stimulus with moving window
    function create_moving_window(words, position){
      var word_list = words.split(' ');
      var stimulus = word_list.map(function(word, index){
        if(index==position){
          return word;
        } else {
          return "";
        }
      }).join(' ')
      return stimulus;
    }

    // create a function for showing the stimulus and collecting the response
    function show_stimulus(position){
      display_element.innerHTML = '<p style="font-size: 24px; font-family:monospace;">' + create_moving_window(trial.words, position) + '</p>';

      jsPsych.pluginAPI.getKeyboardResponse({
        callback_function: after_response,
        valid_responses: [trial.key],
        rt_method: 'performance',
        persist: false,
        allow_held_key: false
      });
    }

    // create a function for handling a response
    function after_response(response_info){
      rt.push(response_info.rt);
      current_word++;
      if(current_word == n_words){
        end_trial();
      } else {
        show_stimulus(current_word);
      }
    }
    
    // create a function to handle ending the trial
    function end_trial(){
      trial_data.rt = JSON.stringify(rt);

      display_element.innerHTML = "";

      jsPsych.finishTrial(trial_data)
    }

    // show the first stimulus
    show_stimulus(current_word);

      // hide word if timing is set
    if (trial.stimulus_duration !== null) {
      jsPsych.pluginAPI.setTimeout(function() {
        display_element.querySelector('<p style="font-size: 24px; font-family:monospace;">' + create_moving_window(trial.words, position) + '</p>').style.visibility = 'hidden';
      }, trial.stimulus_duration);
    }
  };

  return plugin;
})();
```

Additionally if the stimulus_duration is surpassed, I would like to display a message that gives a warning like "You took too long to read that word, so this trial will be ended.", and then I would like the trial to end, and I would like to push to the data a message that the trial was ended early.

How would I accomplish these goals? I would welcome any and all advice. Thank you!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @emilymech ,

You have the right idea here. The trouble is that the `querySelector()` doesn't work by specifying the full set of HTML that you want to select. Instead you have to [specify a valid CSS selector](https://www.javascripttutorial.net/javascript-dom/javascript-queryselector/).

An easy way to do that in your plugin is to add an ID attribute to the stimulus in the `show_stimulus` function, and then select the element based on that ID.

```js
function show_stimulus(position){
      display_element.innerHTML = '<p id="moving-window-stimulus" style="font-size: 24px; font-family:monospace;">' + create_moving_window(trial.words, position) + '</p>';

   // ... //
}
```

```js
 if (trial.stimulus_duration !== null) {
      jsPsych.pluginAPI.setTimeout(function() {
        display_element.querySelector('#moving-window-stimulus').style.visibility = 'hidden';
      }, trial.stimulus_duration);
}
```

One thing to be careful of: you'll need to clear the timeout after each new word is presented. Otherwise, you'll end up in a situation where someone is reading the 3rd word and the time expires from the 1st word and the 3rd word ends up hidden.

You can call `jsPsych.pluginAPI.cancelAllTimeouts()` in the `after_response` function to do this.

> Additionally if the stimulus_duration is surpassed, I would like to display a message that gives a warning like "You took too long to read that word, so this trial will be ended.", and then I would like the trial to end, and I would like to push to the data a message that the trial was ended early.

To do this you'll want to change what happens inside the `setTimeout()` function. Instead of selecting the stimulus and hiding it, you'll want to replace it with the message. 

```js
 if (trial.stimulus_duration !== null) {
      jsPsych.pluginAPI.setTimeout(function() {
        display_element.innerHTML = '<p>You took too long!</p>';
      }, trial.stimulus_duration);
}
```

Then you'll want to add other code inside that function to update the trial data (`trial_data.time_out = true`) and end the trial.

**Reply by emilymech** [reply]
Awesome, thanks so much for all of the help! It's working! :)
